starlette/datastructures.py

```diff
@@ -114,11 +114,18 @@ def replace(self, **kwargs: typing.Any) -> "URL":
             or "hostname" in kwargs
             or "port" in kwargs
         ):
-            hostname = kwargs.pop("hostname", self.hostname)
+            hostname = kwargs.pop("hostname", None)
             port = kwargs.pop("port", self.port)
             username = kwargs.pop("username", self.username)
             password = kwargs.pop("password", self.password)
 
+            if not hostname:
+                netloc = self.netloc
+                _, _, hostname = netloc.rpartition("@")
+
+                if hostname[-1] != "]":
+                    hostname = hostname.rsplit(":", 1)[0]
```

Comment:
I guess the new approach breaks the existing undocumented, untested functionality...

May I propose a test like this:

`assert URL("http://u:p@host/").replace(hostname="bar") == URL("http://u:p@bar/")`

Reply:
Or maybe like this:

`assert URL("http://u:p@host:80").replace(port=88) == URL("http://u:p@host:88")`

Reply:
added, these are passing